*: add kernel argument support

arithx opened the pull request:
Add a new top-level field `kernelArgument` to the configuration specification, a new Ignition stage `kargs`, and distributor documentation.

Closes #1168

nikita-dubrovskii commented on config/v3_3_experimental/types/karg.go:
```diff
@@ -0,0 +1,43 @@
+
+func (k KernelArgument) Validate(c path.ContextPath) (r report.Report) {
+	for i, shouldnt := range k.ShouldntExist {
+		for _, shouldntPart := range strings.Split(string(shouldnt), " ") {
```
wdyt about using `strings.Fields` ?

arithx replied:
Sure.

nikita-dubrovskii commented on config/v3_3_experimental/types/karg.go:
```diff
@@ -0,0 +1,43 @@
+			}
+
+			for _, should := range k.ShouldExist {
+				for _, shouldPart := range strings.Split(string(should), " ") {
```
same question here

nikita-dubrovskii commented on config/v3_3_experimental/types/karg.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/coreos/vcontext/report"
+)
+
+func (k KernelArgument) Validate(c path.ContextPath) (r report.Report) {
```
what should happen when current `cmdline` contains some `foo=boo`and that `karg` is blacklisted with `ShoulndtExist` ?
shouldn't we append `/proc/cmdline` to `ShouldExist` before checking?

arithx replied:
I'm not sure I follow. This is validating that the config isn't contrasting by saying that a specific kernel argument should and should not exist.

nikita-dubrovskii replied:
ok, i misunderstood here - missed that this check happens before modification of `cmdline`.

bgilbert commented on examples/ignition-kargs:
```diff
@@ -0,0 +1,47 @@
+    ;;
+    --should-not-exist)
+    arg="$2"
+    kernelopts="$(echo "$kernelopts" | sed "s,$2,,g")"
```
`--should-not-exist console=ttyS0,115200`

arithx replied:
Going to swap the delimiter to `|` because I don't think any kargs should contain that.

bgilbert commented on examples/ignition-kargs:
```diff
@@ -0,0 +1,47 @@
+if [[ ! "$kernelopts" == "$(grep kernelopts= <$grubcfg)" ]]; then
+    awk "{sub(/kernelopts=.*/,\"$kernelopts\"); print}" $grubcfg > tmp && mv tmp $grubcfg
+
+    reboot
```
At a minimum, if we're not deferring reboots, this should probably be `systemctl reboot`.

arithx commented on config/v3_3_experimental/types/karg_test.go:
```diff
@@ -0,0 +1,66 @@
+	tests := []struct {
+		in  KernelArgument
+		out string
+	}{
```
Verified manually that `ValidateDups` indeed is rejecting in `kernelArguments` for both fields.

arithx commented on docs/distributor-notes.md:
```diff
@@ -0,0 +1,43 @@
+
+## Kernel Arguments
+
+When Ignition is updating kernel arguments it will call out to a binary (defined in `internal/distro/distro.go` and overridable at build-time via overriding the `github.com/coreos/ignition/v2/internal/distro.kargsCmd` build flag). Ignition expects that the binary accepts `--should-exist` & `--should-not-exist` parameters. Should exist operations should append the argument if missing and should no ...
```
Hmm, I don't think it matters either way. We aren't stamping anything from the Ignition side that the `kargs` stage ran (we're expecting that the binary takes care of performing the `no-op`), so the implementation side could choose.

arithx commented on internal/distro/distro.go:
```diff
@@ -62,6 +62,9 @@ var (
 	cryptsetupCmd = "cryptsetup"
 
+	// kargs programs
+	kargsCmd = "ignition-kargs"
```
I renamed it to `ignition-kargs-helper`

arithx commented:
Okay, updated for comments.

Additionally I went ahead and updated `config/merge/merge_test` to use the `v3_3_experimental` config (although I did leave in a `v3_2` for the struct pointers merging).

arithx commented on config/v3_3_experimental/types/karg.go:
```diff
@@ -0,0 +1,64 @@
+	}
+}
+
+func (k KernelArgument) Key() string {
+	if len(k) > 0 {
+		return string(k[0])
+	}
+	return ""
+}
```
@bgilbert While reviewing #1184 I realized that this feels like it might be a problem. Do you have any suggestions for how we could fix this?

bgilbert commented on config/merge/merge_test.go:
```diff
@@ -329,8 +330,8 @@ func TestMerge(t *testing.T) {
 					Luks: []types.Luks{
 						// nested struct pointers, one override
 						{
-							Clevis: &types.Clevis{
```
Let's drop this entire test from `TestMerge`.  It doesn't serve a purpose with the new spec.

bgilbert commented on config/merge/merge_test.go:
```diff
@@ -18,7 +18,8 @@ import (
 
 	"github.com/coreos/ignition/v2/config/util"
-	"github.com/coreos/ignition/v2/config/v3_2/types"
+	v3_2_types "github.com/coreos/ignition/v2/config/v3_2/types"
```
Maybe just `v3_2`?

bgilbert commented on config/merge/merge_test.go:
```diff
@@ -1392,3 +1437,113 @@ func TestMerge(t *testing.T) {
+// pointers. These should not exist but ended up in the Clevis & Luks
+// structs in spec 3.2.0.
+// https://github.com/coreos/ignition/issues/1132
+func TestMerge_3_2_0(t *testing.T) {
```
`TestMergeStructPointers`?  The spec version isn't the relevant part here.  (Same observation for the comment block.)

bgilbert commented on config/merge/merge_test.go:
```diff
@@ -1382,6 +1383,50 @@ func TestMerge(t *testing.T) {
 				{path.New(TAG_CHILD, "storage"), path.New(TAG_RESULT, "storage")},
 			}},
 		},
+		{
```
Comment block describing the test, please.

bgilbert commented on config/v3_3_experimental/types/karg.go:
```diff
@@ -0,0 +1,64 @@
+// Copyright 2021 Red Hat, Inc.
```
nit: maybe rename file to `kargs.go`?

arithx commented on config/v3_3_experimental/types/karg.go:
```diff
@@ -0,0 +1,64 @@
+// done because individual arguments are allowed to have spaces (to
+// allow for position dependent arguments to be set) and MergedKeys only
+// checks for the entire string rather than each individual piece of the
+// argument.
```
Per OOB discussion we felt that this was starting to creep too far in complexity and dropped it in favor of just having entire string matches rejected (via `MergedKeys`). As a result the following things were dropped: the `Validate` function, `config/v3_3_experimental/types/kargs_test.go`, & the new shared error entry.

arithx commented on config/shared/errors/errors.go:
```diff
@@ -82,6 +82,9 @@ var (
 	ErrInvalidInstantiatedUnit = errors.New("invalid systemd instantiated unit")
 
+	// Kernel Argument errors
+	ErrDuplicateKernelArguments = errors.New("kernel argument cannot be present in both shouldExist and shouldNotExist")
```
Dropped as per https://github.com/coreos/ignition/pull/1178#discussion_r604604984

arithx commented:
⏫ Updated ⏫

bgilbert commented on docs/distributor-notes.md:
```diff
@@ -0,0 +1,18 @@
+---
+layout: default
+nav_order: 8
```
Nit: I'd probably put this below Supported Platforms since that's a user-facing doc and this isn't.

bgilbert commented on docs/distributor-notes.md:
```diff
@@ -0,0 +1,18 @@
+
+When Ignition is updating kernel arguments it will call out to a binary (defined in `internal/distro/distro.go` and overridable at build-time via overriding the `github.com/coreos/ignition/v2/internal/distro.kargsCmd` build flag). Ignition expects that the binary accepts `--should-exist` & `--should-not-exist` parameters. Should exist operations should append the argument if missing and should no ...
+
+An example of the binary implementation look at [`examples/ignition-kargs-helper`](https://github.com/coreos/ignition/blob/master/examples/ignition-kargs-helper).
```
Sentence fragment.  "As an example..."?

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,45 @@
+
+while [[ $# -gt 0 ]]
+do
+key="$1"
```
Let's indent the contents of the `while` loop.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,45 @@
+    kernelopts="$(echo "$kernelopts" | sed "s|$2||g")"
+    shift 2
+    ;;
+    *)
```
Bad indent.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,45 @@
+    shift 2
+    ;;
+--should-not-exist)
+    arg="$2"
```
Unused.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,45 @@
+    ;;
+--should-not-exist)
+    arg="$2"
+    kernelopts="$(echo "$kernelopts" | sed "s|$2||g")"
```
`console=ttyS0,115200`, `--should-not-exist 0`.

Or, for that matter, `--should-not-exist kernelopts=`.  It might be cleaner to not carry the `kernelopts=` part in the contents of the variable.

arithx replied:
I've gone ahead and moved the `kernelopts=` out of the variable contents.

As far as the other part, yeah I think that's not great however I'm not sure the best way to approach it without adding a ton of bash complexity. The shell script needs to be able to take something like `--should-not-exist "karg1 karg2"` so doing a simple split of each kernel argument and comparing strings doesn't work. Do you think we should bite the bullet and add bash code to support say splitting the arrays of each and checking if contiguous elements of a split `kernelopts=` array match the array generated by splitting individual `should-exist` & `should-not-exist` entries?

bgilbert replied:
We could use the same trick rdcore does and store `$kernelopts` with a leading and trailing space.  Then we can just `s| $2 | |g`, and drop the leading/trailing spaces when writing `$kernelopts` back to the `grub.cfg`.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,45 @@
+done
+
+# only apply the changes & reboot if changes have been made
+if [[ ! "$kernelopts" == "$(grep kernelopts= <$grubcfg)" ]]; then
```
`!=`?

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,45 @@
+
+# only apply the changes & reboot if changes have been made
+if [[ ! "$kernelopts" == "$(grep kernelopts= <$grubcfg)" ]]; then
+    awk "{sub(/kernelopts=.*/,\"$kernelopts\"); print}" $grubcfg > tmp && mv tmp $grubcfg
```
I'm not sure we can assume that we have `sed` and `awk` in the initrd.  Explicitly installing them from the dracut module would be unnecessary in the common case, so I'm not sure what to do.  Maybe just add a comment noting the dependency?

Would `sed` not work here?

Assumes we're in a writable directory.  Probably better to make `tmp` an absolute path.

arithx replied:
> I'm not sure we can assume that we have `sed` and `awk` in the initrd. Explicitly installing them from the dracut module would be unnecessary in the common case, so I'm not sure what to do. Maybe just add a comment noting the dependency?

:+1:

> Would `sed` not work here?

I spent some time trying to get `sed` to work and couldn't get the incantation exactly right. I can spend some more time if we want to drop the `awk` depdency.

> Assumes we're in a writable directory. Probably better to make `tmp` an absolute path.

Sure, I'll change it to `/run/kargs_tmp`

bgilbert replied:
How about `sed -i "s|^\(kernelopts=\).*|\1$kernelopts|" $grubcfg`?

bgilbert commented on dracut/30ignition/ignition-kargs-helper.sh:
```diff
@@ -0,0 +1,8 @@
+set -euxo pipefail
+
+# Send out a warning message to contact the distribution and exit.
+# Distributions implementing Ignition kargs can either disable the unit
```
They'd disable the unit if they're not implementing kargs, right?

arithx replied:
Yeah, switched up the wording to `Distributions implementing Ignition kargs can override this binary with the actual implementation.`

bgilbert commented on dracut/30ignition/ignition-kargs-helper.sh:
```diff
@@ -0,0 +1,8 @@
+# Send out a warning message to contact the distribution and exit.
+# Distributions implementing Ignition kargs can either disable the unit
+# or override this binary with the actual implementation.
+echo "Ignition kargs requires an additional binary. Please contract your distribution."
```
s/contract/contact/.  Maybe something like `Ignition kernelArguments is not supported by this Linux distribution.\nAsk your distribution to implement ignition-kargs-helper.`?

bgilbert commented on dracut/30ignition/module-setup.sh:
```diff
@@ -59,6 +59,9 @@ install() {
         "/usr/sbin/ignition-setup-user"
 
+    inst_script "$moddir/ignition-kargs-helper.sh" \
+        "/usr/bin/ignition-kargs-helper"
```
`/usr/bin` is inconsistent with `ignition-setup-user`, but maybe less confusing to distro maintainers?

arithx replied:
Switched to `sbin` for consistency.

bgilbert reviewed (commented):
Can we add a test that configs will be rejected if the same value is in `shouldExist` and `shouldNotExist`?

arithx commented:
Updated, will still potentially need another update pending questions in https://github.com/coreos/ignition/pull/1178#discussion_r621845339 & https://github.com/coreos/ignition/pull/1178#discussion_r621846951

bgilbert commented on dracut/30ignition/ignition-kargs-helper.sh:
```diff
@@ -0,0 +1,8 @@
+# Send out a warning message to contact the distribution and exit.
+# Distributions implementing Ignition kargs can override this binary
+# with the actual implementation.
+echo "Ignition kernelArguments is not supported by this Linux distribution.\nAsk your distribution to implement ignition-kargs-helper."
```
Need `echo -e` for the `\n` to work.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,46 @@
+mount -o rw ${bootdev} ${bootmnt}
+grubcfg="${bootmnt}/grub/grub.cfg"
+
+kernelopts="$(grep kernelopts= <$grubcfg | sed \"s,kernelopts=,,g\")"
```
The escaped quotes in the `sed` aren't needed.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,46 @@
+        arg="$2"
+        # don't repeat the arg
+        if [[ ! " ${kernelopts[*]} " =~ " ${arg} " ]]; then
+            kernelopts="$kernelopts $arg "
```
If we have multiple `--should-exist` args, we'll put two spaces between each.

bgilbert commented on config/v3_3_experimental/types/kargs_test.go:
```diff
@@ -0,0 +1,51 @@
+		in  KernelArguments
+		out string
+	}{
+		// Ensure that ValidateWithContext prevents duplicate entires
```
`entries`

arithx commented:
Updated

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,49 @@
+        arg="$2"
+        # don't repeat the arg
+        if [[ ! " ${kernelopts[*]} " =~ " ${arg} " ]]; then
+            kernelopts="$kernelopts $arg"
```
With the trailing whitespace, this should be `$kernelopts$arg ` now.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,49 @@
+mount -o rw ${bootdev} ${bootmnt}
+grubcfg="${bootmnt}/grub/grub.cfg"
+
+kernelopts=" $(grep kernelopts= <$grubcfg | sed s,kernelopts=,,g) "
```
Let's call out the leading/trailing whitespace in a comment.

bgilbert commented on examples/ignition-kargs-helper:
```diff
@@ -0,0 +1,49 @@
+kernelopts="$(echo $kernelopts | sed -e 's,^[[:space:]]*,,' -e 's,[[:space:]]*$,,')"
+
+# only apply the changes & reboot if changes have been made
+if [[ "$kernelopts" != "$(grep kernelopts= <$grubcfg | sed s,kernelopts=,,g)" ]]; then
```
The RHS is sufficiently complicated now that it might be worth just doing `orig_kernelopts="$kernelopts"` at the top of the script.
